File: kina-examples/src/main/java/kina/examples/java/GroupingCellWithMongoDB.java
/*
 * Copyright 2014, Luca Rosellini.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kina.examples.java;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import kina.config.MongoConfigFactory;
import kina.config.MongoKinaConfig;
import kina.context.MongoKinaContext;
import kina.entity.Cells;
import kina.entity.MongoCell;
import kina.rdd.mongodb.MongoCellRDD;
import kina.testutils.ContextProperties;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;
import scala.Tuple2;

/**
 * Created by rcrespo on 25/06/14.
 */
public final class GroupingCellWithMongoDB {


    private GroupingCellWithMongoDB() {
    }


    public static void main(String[] args) {
        doMain(args);
    }


    public static void doMain(String[] args) {
        String job = "scala:groupingCellWithMongoDB";

        String host = "localhost:27017";

        String database = "book";
        String inputCollection = "input";
        String outputCollection = "output";


        // Creating the Kina Context where args are Spark Master and Job Name
        ContextProperties p = new ContextProperties(args);
	    MongoKinaContext deepContext = new MongoKinaContext(p.getCluster(), job, p.getSparkHome(),
                p.getJars());


        MongoKinaConfig<Cells> inputConfigEntity =
				        MongoConfigFactory.createMongoDB().host(host).database(database).collection(inputCollection).initialize();

        JavaRDD<Cells> inputRDDEntity = deepContext.mongoJavaRDD(inputConfigEntity);


        JavaRDD<String> words =inputRDDEntity.flatMap(new FlatMapFunction<Cells, String>() {
            @Override
            public Iterable<String> call(Cells cells) throws Exception {

                List<String> words = new ArrayList<>();
                for (Cells canto : (List<Cells>)cells.getCellByName("cantos").getCellValue() ){
                    words.addAll(Arrays.asList( ((String)canto.getCellByName("text").getCellValue()).split(" ")));
                }
                return words;
            }
        });


        JavaPairRDD<String, Integer> wordCount = words.mapToPair(new PairFunction<String, String, Integer>() {
            @Override
            public Tuple2<String, Integer> call(String s) throws Exception {
                return new Tuple2<String, Integer>(s,1);
            }
        });


        JavaPairRDD<String, Integer>  wordCountReduced = wordCount.reduceByKey(new Function2<Integer, Integer, Integer>() {
            @Override
            public Integer call(Integer integer, Integer integer2) throws Exception {
                return integer + integer2;
            }
        });

        JavaRDD<Cells>  outputRDD =  wordCountReduced.map(new Function<Tuple2<String, Integer>, Cells>() {
            @Override
            public Cells call(Tuple2<String, Integer> stringIntegerTuple2) throws Exception {
                return new Cells(MongoCell.create("word", stringIntegerTuple2._1()) , MongoCell.create("count", stringIntegerTuple2._2()));
            }
        });


        MongoKinaConfig<Cells> outputConfigEntity =
				        MongoConfigFactory.createMongoDB().host(host).database(database).collection(outputCollection).initialize();

        MongoCellRDD.saveCell(outputRDD.rdd(), outputConfigEntity);

        deepContext.stop();
    }
}

File: kina-examples/src/main/scala/kina/examples/scala/GroupingCellWithMongoDB.scala
/*
 * Copyright 2014, Luca Rosellini.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package kina.examples.scala


import kina.config._
import kina.context.MongoKinaContext
import kina.entity.{Cells, MongoCell}
import kina.rdd.mongodb.MongoCellRDD
import kina.testutils.ContextProperties
import org.apache.spark.rdd.RDD
import org.apache.spark.SparkContext._

/**
 * Created by rcrespo on 25/06/14.
 */
final object GroupingCellWithMongoDB {
  def main(args: Array[String]) {
    doMain(args)
  }

  def doMain(args: Array[String]) {
    val job: String = "scala:groupingCellWithMongoDB"
    val host: String = "localhost:27017"
    val database: String = "book"
    val inputCollection: String = "input"
    val outputCollection: String = "output"
    val p: ContextProperties = new ContextProperties(args)
    val kinaContext: MongoKinaContext = new MongoKinaContext(p.getCluster, job, p.getSparkHome, p.getJars)
    val inputConfigEntity: MongoKinaConfig[Cells] = MongoConfigFactory.createMongoDB.host(host).database(database).collection(inputCollection).initialize
    val inputRDDEntity: RDD[Cells] = kinaContext.mongoRDD(inputConfigEntity)


    val words: RDD[String] = inputRDDEntity flatMap {
      e: Cells => (for (canto : Cells<- e.getCellByName("cantos").getCellValue().asInstanceOf[List[Cells]])
      yield canto.getCellByName("text").getCellValue.asInstanceOf[String].split(" ")).flatten
    }


    val wordCount : RDD[(String, Integer)] = words map { s:String => (s,1) }

    val wordCountReduced  = wordCount reduceByKey { (a,b) => a + b }

    val outputRDD = wordCountReduced map { e:(String, Integer) => new Cells(MongoCell.create("word", e._1), MongoCell.create("count",e._2))  }

    val outputConfigEntity: MongoKinaConfig[Cells] =
      MongoConfigFactory.createMongoDB().host(host).database(database).collection(outputCollection).initialize

    MongoCellRDD.saveCell(outputRDD, outputConfigEntity)

    kinaContext.stop
  }
}


